src/table/iterator.rs

```diff
@@ -0,0 +1,170 @@
+use super::builder::{Header, HEADER_SIZE};
+use super::{Block, TableInner};
+use crate::util::{self, KeyComparator, COMPARATOR};
+use crate::value::Value;
+use bytes::Bytes;
+use std::sync::Arc;
+
+#[derive(Clone, Debug)]
+pub enum IteratorError {
+    NoError,
+    EOF,
+    // TODO: As we need to clone Error from block iterator to table iterator,
+    // we had to save `crate::Error` as String. In the future, we could let all
+    // seek-related function to return a `Result<()>` instead of saving an
+    // error inside struct.
+    Error(String),
+}
+
+impl IteratorError {
+    pub fn is_err(&self) -> bool {
+        match self {
+            IteratorError::NoError => false,
+            _ => true,
+        }
+    }
+
+    pub fn is_eof(&self) -> bool {
+        match self {
+            IteratorError::EOF => true,
+            _ => false,
+        }
+    }
+}
+
+enum SeekPos {
+    Origin,
+    Current,
+}
+
+/// Block iterator iterates on an SST block
+// TODO: support custom comparator
+struct BlockIterator {
+    idx: isize,
+    base_key: Bytes,
+    key: Bytes,
+    val: Bytes,
+    data: Bytes,
+    // TODO: use `&'a Block` if possible
+    block: Arc<Block>,
+    perv_overlap: u16,
+    err: IteratorError,
+}
+
+impl BlockIterator {
+    pub fn new(block: Arc<Block>) -> Self {
+        let data = block.data.slice(..block.entries_index_start);
+        Self {
+            block,
+            err: IteratorError::NoError,
+            base_key: Bytes::new(),
+            key: Bytes::new(),
+            val: Bytes::new(),
+            data,
+            perv_overlap: 0,
+            idx: 0,
+        }
+    }
+
+    #[inline]
+    fn entry_offsets(&self) -> &[u32] {
+        &self.block.entry_offsets
+    }
+
+    fn set_idx(&mut self, i: isize) {
+        self.idx = i;
+        if i >= self.entry_offsets().len() as isize || i < 0 {
+            self.err = IteratorError::EOF;
+            return;
+        }
+
+        self.err = IteratorError::NoError;
+
+        // if base key is not available, retrieve it from first key-value pair in block
+        if self.base_key.is_empty() {
```

Comment:
use `Option<>`?

Reply:
This was intended, as originally I thought these two structs will take different spaces. Just now I found that `Option<Iterator>` takes the same space as current implementation. I'll change to `Option<>`

<img width="984" alt="image" src="https://user-images.githubusercontent.com/4198311/96564028-3aef9780-12f5-11eb-9cdb-9cfdcb7a4735.png">

Reply:
FYI, now it takes 32 bytes.